sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisReader.java

```diff
@@ -142,4 +159,28 @@ public Instant getWatermark() {
     return source;
   }
 
+  /**
+   * Returns total size of all records that remain in Kinesis stream after current watermark.
+   * When currently processed record is not further behind than {@link #upToDateThreshold}
+   * then this method returns 0.
+   */
+  @Override
+  public long getTotalBacklogBytes() {
+    Instant watermark = getWatermark();
+    if (watermark.plus(upToDateThreshold).isAfterNow()) {
+      return 0L;
```

Comment:
@pawel-kaczmarczyk I don't understand why cannot we simply return the actual number of backlog bytes? It should be up to the execution service to decide whether that number is high enough that scaling should occur

Reply:
We can simply return the backlog bytes and it works well in most cases. However in our case we observed unnecessary scaling during short spikes. Also there were issues when there were not so many events in the stream backlog but Kinesis was slow to return them. In such situation events are lagging behind only for a minute or two but runner may still decide to scale up.

Reply:
Fair enough.
Can you please write that explanation in the comment in the code? Otherwise it's hard to understand why it is here.
Also, I'm not big fan of the "upToDataThreshold", but can't really come up with anything much better

Reply:
This sounds like something that should be fixed in the Dataflow Runner's streaming autoscaling logic (which, conveniently, is @rangadi's realm :) ), rather than in KinesisIO - though I'm ok keeping it as a temporary workaround while Dataflow Runner is being fixed. Could you file a JIRA explaining the behavior you observed from Dataflow runner's autoscaling, and link it from here, with a TODO to remove the knob?

Also @rangadi can you confirm that according to your understanding this knob is in fact conceptually a good workaround for the poor behavior of streaming autoscaling in this case?

Reply:
Sort of yes. In Dataflow autoscaling, some of the important parameters are not tunable. E.g. it upscales if the estimated time for remaining work is longer than 15 seconds. Not all pipelines might be this sensitive.. also in this case the actual backlog is reported by another aggregator of metrics which seems to update for every minute. 

This is a reasonable work around. I'm not a fan of the name of the method either, but don't have much better suggestion.